enhancements/machine-config/manage-boot-images.md

```diff
@@ -0,0 +1,494 @@
+
+#### Opt-in Mechanism
+
+This proposal will introduce a discriminated union in [operator types](https://github.com/openshift/api/blob/master/operator/v1/types_machineconfiguration.go) for the MCO, `ManagedBootImageConfig` which has two fields:
+
+- `Mode` This is a string enum which can have three values:
+  - `Enabled` - All `Machinesets` will be enrolled for boot image updates.
+  - `CustomConfig` - `Machinesets` matched with the label selector will be enrolled for boot image updates.
+  - `Disabled` - No `Machinesets` will be enrolled for boot image updates.
+- `CustomConfig` This is struct which encloses a label selector that will be used by machineset objects to opt-in. 
+
+Here are some YAML examples that describes operators in each of these modes:
+##### Enabled
+```
+apiVersion: operator.openshift.io/v1
+kind: MachineConfiguration
+metadata:
+  name: default
+  labels:
+spec:
+  managedBootImageConfig:
+    mode: Enabled
+```
+##### Disabled
+```
+apiVersion: operator.openshift.io/v1
+kind: MachineConfiguration
+metadata:
+  name: default
+  labels:
+spec:
+  managedBootImageConfig:
+    mode: Disabled
+```
+##### MatchSelector
+```
+apiVersion: operator.openshift.io/v1
+kind: MachineConfiguration
+metadata:
+  name: default
+  labels:
+spec:
+  managedBootImageConfig:
+    mode: CustomConfig
+    CustomConfig:
+      machineSetSelector:
+        matchLabels:
+          machineconfiguration.openshift.io/mco-managed-machineset: ""        
+```
+Note: While in this mode, the label added to the selector will have to be added to the `machineset` object.
+
+A [ValidatingAdmissionPolicy](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/) will be implemented via an MCO manifest that will restrict updating the `ManagedBootImageConfig` object to only supported platforms(initially, just GCP). This will be updated as we phase in support for other platforms. Here is a sample policy that would do this:
+
+```
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "managed-bootimages-platform-check"
+spec:
+  failurePolicy: Fail
+  paramKind:
+    apiVersion: config.openshift.io/v1
+    kind: Infrastructure
+  matchConstraints:
+    resourceRules:
+    - apiGroups:   ["operator"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["MachineConfiguration"]
+  validations:
+    - expression: "has(object.spec.MachineBootImageConfig) && param.status.platformStatus.Type != `GCP`"
+      message: "This feature is only supported on these platforms: GCP"
+```
+This would need an accompanying binding:
+```
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: ValidatingAdmissionPolicyBinding
+metadata:
+  name: "managed-bootimages-platform-check-binding"
+spec:
+  policyName: "managed-bootimages-platform-check"
+  validationActions: [Deny]
+  paramRef:
+    name: "cluster"
+    namespace: "default"
+```
+#### Tracking boot image history
```

Comment:
Do we have an o/api PR for this?

Reply:
We do not, I do not envision this work happening in the short term. At the very least, I figured we could outline it here and based on team capacity, we could work on adding this capability.

Reply:
Ok, can you make it clear in the enhancement then that this is just an idea and that it's future work. I wouldn't want to approve the EP saying that this API is good to go, when we haven't been through the API review process